service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphGenericRepository.java

```diff
@@ -242,15 +333,90 @@ public static String getVertexStringPropertyOrNull(final Vertex vertex, final St
         return null;
     }
 
-    abstract String computeId(E entity);
+    public E getByZoneAndIdentifier(final ZoneEntity zone, final String identifier) {
+        try {
+            GraphTraversal<Vertex, Vertex> traversal = getGraph().traversal().V().has(ZONE_ID_KEY, zone.getName())
+                    .has(getEntityIdKey(), identifier);
+            if (!traversal.hasNext()) {
+                return null;
+            }
+            Vertex vertex = traversal.next();
+            E entity = vertexToEntity(vertex);
+            searchAttributes(entity, vertex);
+
+            // There should be only one entity with a given entity id.
+            Assert.isTrue(!traversal.hasNext(),
+                    String.format("There are two entities with the same %s.", getEntityIdKey()));
+            return entity;
+        } finally {
+            getGraph().tx().commit();
+        }
+    }
+
+    public E getByZoneAndIdentifierAndScopes(final ZoneEntity zone, final String identifier,
+            final Set<Attribute> scopes) {
+        try {
+            GraphTraversal<Vertex, Vertex> traversal = getGraph().traversal().V().has(ZONE_ID_KEY, zone.getName())
+                    .has(getEntityIdKey(), identifier);
+            if (!traversal.hasNext()) {
+                return null;
+            }
+            Vertex vertex = traversal.next();
+            E entity = vertexToEntity(vertex);
+            searchAttributesWithScopes(entity, vertex, scopes);
+
+            // There should be only one entity with a given entity id.
+            Assert.isTrue(!traversal.hasNext(),
+                    String.format("There are two entities with the same %s.", getEntityIdKey()));
+            return entity;
+        } finally {
+            getGraph().tx().commit();
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public void searchAttributes(final E entity, final Vertex vertex) {
```

Comment:
is there a unit test which directly tests this method ?  at this level we can test cases which are not in the Resource/Subject repository tests
- vertex with outbound edges to 2 other vertices  (subject or resource with 2 parents)
- others ?